builtin/providers/openstack/resource_openstack_blockstorage_volume_attach_v2.go

```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return err
+	}
+
+	var attachmentId string
+	for _, attachment := range volume.Attachments {
+		if instanceId != "" && instanceId == attachment.ServerID {
+			attachmentId = attachment.AttachmentID
+		}
+
+		if hostName != "" && hostName == attachment.HostName {
+			attachmentId = attachment.AttachmentID
+		}
+	}
+
+	if attachmentId == "" {
+		return fmt.Errorf("Unable to determine attachment ID.")
+	}
+
+	// The ID must be a combination of the volume and attachment ID
+	// in order to import attachments.
+	id := fmt.Sprintf("%s/%s", volumeId, attachmentId)
+	d.SetId(id)
+
+	return resourceBlockStorageVolumeAttachV2Read(d, meta)
+}
+
+func resourceBlockStorageVolumeAttachV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.blockStorageV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+
+	volumeId, attachmentId, err := blockStorageVolumeAttachV2ParseId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	volume, err := volumes.Get(client, volumeId).Extract()
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Retrieved volume %s: %#v", d.Id(), volume)
+
+	var attachment volumes.Attachment
+	for _, v := range volume.Attachments {
+		if attachmentId == v.AttachmentID {
+			attachment = v
+		}
+	}
+
+	log.Printf("[DEBUG] Retrieved volume attachment: %#v", attachment)
+
+	d.Set("volume_id", volumeId)
+	d.Set("attachment_id", attachmentId)
+	d.Set("device", attachment.Device)
+	d.Set("instance_id", attachment.ServerID)
+	d.Set("host_name", attachment.HostName)
+
+	return nil
+}
+
+func resourceBlockStorageVolumeAttachV2Delete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.blockStorageV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+
+	volumeId, attachmentId, err := blockStorageVolumeAttachV2ParseId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	detachOpts := volumeactions.DetachOpts{
+		AttachmentID: attachmentId,
+	}
+
+	log.Printf("[DEBUG] Detachment Options: %#v", detachOpts)
+
+	if err := volumeactions.Detach(client, volumeId, detachOpts).ExtractErr(); err != nil {
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"in-use", "attaching", "detaching"},
+		Target:     []string{"available"},
+		Refresh:    VolumeV2StateRefreshFunc(client, volumeId),
+		Timeout:    10 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for volume (%s) to become available: %s", volumeId, err)
+	}
+
+	return nil
+}
+
+func blockStorageVolumeAttachV2AttachMode(v string) (volumeactions.AttachMode, error) {
+	var attachMode volumeactions.AttachMode
+	var attachError error
+	switch v {
+	case "":
```

Comment:
Well if we're going to be honest with each other, I gotta say _IMO_, I'm not a fan either :)

So this is a legitimate `case` statement that is kind of equivalent to:

```go
switch v {
case "":
  nil
case ...:
```

This is because there could be three valid options:

1. Don't specify and let OpenStack apply the default (which is RW, but that _might_ be different depending on the cloud).
2. RW
3. RO

I'll fix this up and make the first case more explicit rather than obscure. I'd rather not enforce a default at the schema level since different OpenStack clouds _could_ be doing something different than the standard default (which is something else I don't agree with, but I respect the right do to it :)